quacs-rs: use deterministic iteration in test_freshman_cs()

johnnyapol opened the pull request:
The ordering of a map key iteration is undefined and therefore
this test fails if the ordering is changed.

This fixes test failures in https://github.com/quacs/quacs/pull/527

io12 commented:
Another potential solution with less boilerplate: change `phf_map!` to [`phf_ordered_map!`](https://docs.rs/phf/0.9.0/phf/macro.phf_ordered_map.html).

io12 commented:
Changing it from a map to an array of tuples might be a good idea too, because then the special case indices (0 and 9) don't need to be changed.

johnnyapol commented:
Unfortunately, the conversion to phf_ordered_map is not very straightforward as this requires some more changes in the SemesterData struct (which is hard-coded to use a phf::Map).

I've thought about this and the only real way this could be done is to make SemesterData take a custom trait that provides a common get() function implementation for the two maps. However, we now need to update all references of SemesterData to take this trait into account. Needless to say, it gets very messy very quickly (and adds a lot of typing boilerplate).

johnnyapol commented:
Another hacky solution would be to right a conversion from phf::ordered_map to phf::map; however, the internal structure is not well defined according to the documentation. This means we'd need to go the macro/code-gen route to duplicate things and frankly I'm not equipped enough to come up with that.

johnnyapol commented:
FWIW, this was the trait I came up with
`trait Table<K,V> {
    fn get<T: ?Sized>(&self, key: &T) -> Option<&V>
    where T: Eq + PhfHash,
    K: PhfBorrow<T>;
}

impl<K,V> Table<K,V> for phf::Map<K,V> {
    fn get<T: ?Sized>(&self, key: &T) -> Option<&V>
        where T: Eq + PhfHash,
        K: PhfBorrow<T> {
        return self.get(key);
    }
}

impl<K,V> Table<K,V> for phf::OrderedMap<K,V> {
    fn get<T: ?Sized>(&self, key: &T) -> Option<&V>
        where T: Eq + PhfHash,
        K: PhfBorrow<T> {
        return self.get(key);
    }
}`

However, by the time we've fixed up SemesterData to use it (+ made the external changes), we've added as much code.

A third solution could be dynamic dispatch but then we incur that overhead at runtime for code that's only ever used in tests.

io12 commented:
Another potential solution (untested, but probably works):

```rust
let iteration_order = {
    let mut v = crn_times.keys.iter().collect::<Vec<u32>>();
    v.sort();
    v
};
```

If we want to take itertools as a test dependency that has `.sorted()`.

johnnyapol commented:
At first I tried sorting it but the sorted order fails the test case.

io12 commented:
> At first I tried sorting it but the sorted order fails the test case.

You'd need to change the `i == 0 || i == 9` part to whatever the new indices are. Changing it to `false` and using the print output to add the cases is how I found them originally.

io12 commented:
It looks like `i == 0 || i == 1 || i == 2` works.

io12 commented:
This seems to work

```diff
diff --git a/site/src/quacs-rs/src/test_context.rs b/site/src/quacs-rs/src/test_context.rs
index b58494d..532aecf 100644
--- a/site/src/quacs-rs/src/test_context.rs
+++ b/site/src/quacs-rs/src/test_context.rs
@@ -276,8 +276,15 @@ fn test_freshman_cs() {
     // No sections conflict
     assert!(crn_times.keys().all(|crn| !ctx.is_in_conflict(*crn)));
 
+    // Get a sorted list of CRNs so iteration order is deterministic
+    let sorted_crns = {
+        let mut crns = crn_times.keys().copied().collect::<Vec<u32>>();
+        crns.sort();
+        crns
+    };
+
     // Select all sections, recomputing all conflicts after each section
-    for (i, crn) in crn_times.keys().enumerate() {
+    for (i, crn) in sorted_crns.iter().enumerate() {
         ctx.set_selected(*crn, true);
 
         // Compute conflicts
@@ -288,7 +295,7 @@ fn test_freshman_cs() {
 
         println!("{}: {}", i, crn);
 
-        if i == 0 || i == 9 {
+        if i == 0 || i == 1 || i == 2 {
             // No sections conflict
             assert!(conflicts.iter().all(|b| !b));
         } else {
```

johnnyapol commented:
Yeah I was playing with it and came to the same conclusion. Latest push has the fix

johnnyapol commented:
@benjaminrsherman this should be ready now unless any other objections are raised
